src/Features/CSharp/Portable/UseExpressionBody/CodeRefactoringProviders/AbstractUseExpressionBodyCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Options;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseExpressionBody
+{
+    internal abstract class AbstractUseExpressionBodyCodeRefactoringProvider<TDeclaration> :
+        CodeRefactoringProvider
+        where TDeclaration : SyntaxNode
+    {
+        private readonly AbstractUseExpressionBodyHelper<TDeclaration> _helper;
+
+        protected AbstractUseExpressionBodyCodeRefactoringProvider(
+            AbstractUseExpressionBodyHelper<TDeclaration> helper)
+        {
+            _helper = helper;
+        }
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            if (context.Span.Length > 0)
+            {
+                return;
+            }
+
+            var position = context.Span.Start;
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var node = root.FindToken(position).Parent;
+            if (node == null)
+            {
+                return;
+            }
+
+            var containingLambda = node.FirstAncestorOrSelf<LambdaExpressionSyntax>();
```

Comment:
❔ Does this cover other forms (`delegate { stuff... }`)? If not, should it?

❔ What about local functions? I don't think we actually care since it's not possible for something containing a local function to use the expression-bodied form.